algebird-core/src/main/scala/com/twitter/algebird/AdaptiveCache.scala

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2015 Twitter, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package com.twitter.algebird
+
+/**
+ * @author Avi Bryant
+ */
+
+import collection.mutable.{ ListBuffer, HashMap }
+import ref.{ SoftReference, ReferenceQueue }
+
+/**
+ * This is a summing cache whose goal is to grow until we run out of memory,
+ * at which point it clears itself and stops growing.
+ * Note that we can lose the values in this cache at any point;
+ * we don't put anything here we care about.
+ */
+class SentinelCache[K, V](implicit sgv: Semigroup[V]) {
```

Comment:
the name here bothers me a bit because from the user's standpoint the only useful thing you get is size, right? I wonder why you are summing in here at all? And aren't we summing twice?